src/workerd/api/streams/readable-source.c++

```diff
@@ -0,0 +1,817 @@
+                {"gzip compressed stream ended prematurely"_kj,
+                  "Gzip compressed stream ended prematurely."_kj},
+                {"gzip decompression failed"_kj, "Gzip decompression failed."},
+                {"brotli state allocation failed"_kj, "Brotli state allocation failed."},
+                {"invalid brotli window size"_kj, "Invalid brotli window size."},
+                {"invalid brotli compression level"_kj, "Invalid brotli compression level."},
+                {"brotli window size too big"_kj, "Brotli window size too big."},
+                {"brotli decompression failed"_kj, "Brotli decompression failed."},
+                {"brotli compression failed"_kj, "Brotli compression failed."},
+                {"brotli compressed stream ended prematurely"_kj,
+                  "Brotli compressed stream ended prematurely."},
+              })) {
+        kj::throwFatalException(kj::mv(translated));
+      } else {
+        kj::throwFatalException(kj::mv(exception));
+      }
+    }
+  }
+
+  kj::Maybe<Tee> tryTee(kj::Maybe<size_t> maybeLimit) override {
+    // Note that if we haven't called read() yet, then the inner stream is still
+    // in its original encoding. If read() has been called, however, then the inner
+    // stream will be wrapped and will be in identity encoding.
+    static constexpr size_t kMax = kj::maxValue;
+    auto& inner = KJ_ASSERT_NONNULL(getState().tryGet<kj::Own<kj::AsyncInputStream>>());
+    auto tee = kj::newTee(kj::mv(inner), maybeLimit.orDefault(kMax));
+    return Tee{
+      .branch1 =
+          kj::heap<EncodedAsyncInputStream>(wrapTeeBranch(kj::mv(tee.branches[0])), getEncoding()),
+      .branch2 =
+          kj::heap<EncodedAsyncInputStream>(wrapTeeBranch(kj::mv(tee.branches[1])), getEncoding()),
+    };
+  }
+
+  kj::AsyncInputStream& ensureIdentityEncoding(kj::Own<kj::AsyncInputStream>& inner) override {
+    auto encoding = getEncoding();
+    if (encoding == rpc::StreamEncoding::IDENTITY) return *inner;
+    setEncoding(rpc::StreamEncoding::IDENTITY);
+    return setStream(wrap(encoding, kj::mv(inner)));
+  }
+
+ private:
+  static kj::Own<kj::AsyncInputStream> wrap(
+      rpc::StreamEncoding encoding, kj::Own<kj::AsyncInputStream> inner) {
+    switch (encoding) {
+      case rpc::StreamEncoding::IDENTITY: {
+        return kj::mv(inner);
+      }
+      case rpc::StreamEncoding::GZIP: {
+        return kj::heap<kj::GzipAsyncInputStream>(*inner).attach(kj::mv(inner));
+      }
+      case rpc::StreamEncoding::BROTLI: {
+        return kj::heap<kj::BrotliAsyncInputStream>(*inner).attach(kj::mv(inner));
+      }
+    }
+    KJ_UNREACHABLE;
+  }
+};
+
+}  // namespace
+
+kj::Own<ReadableStreamSource> newReadableStreamSourceFromBytes(
+    kj::ArrayPtr<const kj::byte> bytes, kj::Maybe<kj::Own<void>> maybeBacking) {
+  KJ_IF_SOME(backing, maybeBacking) {
+    auto inner = newMemoryInputStream(bytes);
+    return newReadableStreamSource(inner.attach(kj::mv(backing)));
```

Comment:
I see `MemoryInputStream` already supports a backing buffer parameter. Why don't we use that? I guess because `ReadableStreamSourceImpl` never calls `tryTee()` on the inner stream, so this lets us avoid the extra Refcounted overhead?

If that's so, comments would helpful.